crates/stackable-operator/src/validation.rs

```diff
@@ -26,121 +27,202 @@ const RFC_1123_SUBDOMAIN_MAX_LENGTH: usize = 253;
+        .expect("failed to compile Kerberos realm name regex")
+});
 
-/// Returns a formatted error message for regex violations.
-///
-/// # Arguments
-///
-/// * `msg` - this is the main error message to return
-/// * `fmt` - this is the regular expression that did not match the input
-/// * `examples` - are optional well, formed examples that would match the regex
-fn regex_error(msg: &str, fmt: &str, examples: &[&str]) -> String {
-    if examples.is_empty() {
-        return format!("{msg} (regex used for validation is '{fmt}')");
-    }
+type Result<T = (), E = Errors> = std::result::Result<T, E>;
+
+/// A collection of errors discovered during validation.
+#[derive(Debug)]
+pub struct Errors(Vec<Error>);
 
-    let mut msg = msg.to_string();
-    msg.push_str(" (e.g. ");
-    for (i, example) in examples.iter().enumerate() {
-        if i > 0 {
-            msg.push_str(" or ");
+impl Display for Errors {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (i, error) in self.0.iter().enumerate() {
+            let prefix = match i {
+                0 => "",
+                _ => ", ",
+            };
+            write!(f, "{prefix}{error}")?;
         }
-        msg.push('\'');
-        msg.push_str(example);
-        msg.push_str("', ");
+        Ok(())
     }
+}
+impl std::error::Error for Errors {}
+
+/// A single validation error.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(transparent)]
+    Regex { source: RegexError },
+    #[snafu(display("input is {length} bytes long but must be no more than {max_length}"))]
+    TooLong { length: usize, max_length: usize },
+}
 
-    msg.push_str("regex used for validation is '");
-    msg.push_str(fmt);
-    msg.push_str("')");
-    msg
+#[derive(Debug)]
+pub struct RegexError {
+    /// The primary error message.
+    msg: &'static str,
+    /// The regex that the input must match.
+    regex: &'static str,
+    /// Examples of valid inputs (if non-empty).
+    examples: &'static [&'static str],
 }
 
-/// Tests for a string that conforms to the definition of a subdomain in DNS (RFC 1123).
-pub fn is_rfc_1123_subdomain(value: &str) -> Result<(), Vec<String>> {
-    let mut errors = vec![];
-    if value.len() > RFC_1123_SUBDOMAIN_MAX_LENGTH {
-        errors.push(max_len_error(RFC_1123_SUBDOMAIN_MAX_LENGTH))
+impl Display for RegexError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self {
+            msg,
+            regex,
+            examples,
+        } = self;
+        write!(f, "{msg} (")?;
+        if !examples.is_empty() {
+            for (i, example) in examples.iter().enumerate() {
+                let prefix = match i {
+                    0 => "e.g.",
+                    _ => "or",
+                };
+                write!(f, "{prefix} {example:?}, ")?;
+            }
+        }
+        write!(f, "regex used for validation is {regex:?})")
```

Comment:
style: Give this code some breathing room.

```suggestion
        write!(f, "{msg} (")?;

        if !examples.is_empty() {
            for (i, example) in examples.iter().enumerate() {
                let prefix = match i {
                    0 => "e.g.",
                    _ => "or",
                };
                write!(f, "{prefix} {example:?}, ")?;
            }
        }

        write!(f, "regex used for validation is {regex:?})")
```

Reply:
All that oxidation impacts the code's ability to breathe /s